internal/dot/graph.go

```diff
@@ -20,53 +20,166 @@
 
 package dot
 
-import "fmt"
+import (
+	"fmt"
+	"reflect"
+)
 
 // Ctor encodes a constructor provided to the container for the DOT graph.
 type Ctor struct {
-	Name    string
-	Package string
-	File    string
-	Line    int
-	Params  []*Node
-	Results []*Node
+	Name       string
+	Package    string
+	File       string
+	Line       int
+	Params     []*Param
+	GroupParam []*Group
+	Results    []*Result
+}
+
+// Param is a parameter node in the graph.
+type Param struct {
+	*Node
+	Optional bool
+}
+
+// Result is a result node in the graph.
+type Result struct {
+	*Node
+	GroupIndex int
+}
+
+// Group is a group node in the graph.
+type Group struct {
+	Type    reflect.Type
+	Group   string
+	Results []*Result
 }
 
 // Graph is the DOT-format graph in a Container.
 type Graph struct {
-	Ctors []*Ctor
+	Ctors  []*Ctor
+	Groups map[key]*Group
 }
 
-// Node is a single node in a graph.
+// Node is a single node in a graph and is embedded into Params and Results.
 type Node struct {
-	Type     string
-	Name     string
-	Optional bool
-	Group    string
+	Type  reflect.Type
+	Name  string
+	Group string
+}
+
+type key struct {
+	t reflect.Type
+	g string
+}
+
+// NewGraph creates an empty graph.
+func NewGraph() *Graph {
+	return &Graph{
+		Groups: make(map[key]*Group),
+	}
+}
+
+// NewGroup creates a new group with information in the key.
+func NewGroup(k key) *Group {
+	return &Group{
+		Type:  k.t,
+		Group: k.g,
+	}
+}
+
+// AddCtor adds the constructor with paramList and resultList into the graph.
+func (dg *Graph) AddCtor(c *Ctor, paramList []*Param, resultList []*Result) {
+	var params []*Param
+	var groupParam []*Group
+
+	// Loop through the paramList to separate them into regular params and
+	// grouped params. For grouped params, we use getGroup to find the actual
+	// group.
+	for _, param := range paramList {
+		if param.Group != "" {
+			k := key{t: param.Type.Elem(), g: param.Group}
+			group := dg.getGroup(k)
+			groupParam = append(groupParam, group)
+		} else {
+			params = append(params, param)
+		}
```

Comment:
nit: Go encourages reducing nesting of primary logic by handling "special" cases first.

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

So this would be,

```
if param.Group == "" {
  // Not a value group.
  params = append(params, param)
  continue
}

k := key{...}
...
groupParams = append(groupParams, group)
```

Reply:
thanks for including the link. I've seen that elsewhere but I didn't know it was a Go convention. Fixed in next commit!